libcore/FileOperations.vala

```diff
@@ -0,0 +1,233 @@
+            base (parent_window);
+            if (trash_dirs != null) {
+                this.trash_dirs = (owned) trash_dirs;
+            } else {
+                this.trash_dirs = new GLib.List<GLib.File> ();
+                this.trash_dirs.prepend (GLib.File.new_for_uri ("trash:"));
+            }
+        }
+
+        /* Only called if confirmation known to be required - do not second guess */
+        private bool confirm_empty_trash () {
+            unowned GLib.File? first_dir = trash_dirs.nth_data (0);
+            if (first_dir != null) {
+                unowned string primary = null;
+                unowned string secondary = null;
+                if (first_dir.has_uri_scheme ("trash")) {
+                    /* Empty all trash */
+                    primary = _("Permanently delete all items from Trash?");
+                    secondary = _("All items in all trash directories, including those on any mounted external drives, will be permanently deleted.");
+                } else {
+                    /* Empty trash on a particular mounted volume */
+                    primary = _("Permanently delete all items from Trash on this mount?");
+                    secondary = _("All items in the trash on this mount, will be permanently deleted.");
+                }
+
+                var message_dialog = new Granite.MessageDialog.with_image_from_icon_name (
+                    primary,
+                    secondary,
+                    "dialog-warning",
+                    Gtk.ButtonsType.CANCEL
+                );
+
+                message_dialog.transient_for = parent_window;
+                unowned Gtk.Widget empty_button = message_dialog.add_button (EMPTY_TRASH, Gtk.ResponseType.YES);
+                empty_button.get_style_context ().add_class (Gtk.STYLE_CLASS_DESTRUCTIVE_ACTION);
+                Gtk.ResponseType response = (Gtk.ResponseType) message_dialog.run ();
+                message_dialog.destroy ();
+
+                return response == Gtk.ResponseType.YES;
+            }
+
+            return true;
+        }
+
+        private async void delete_trash_file (GLib.File file, bool delete_file = true, bool delete_children = true) {
+            if (aborted ()) {
+                return;
+            }
+
+            if (delete_children) {
+                try {
+                    const string ATTRIBUTES = GLib.FileAttribute.STANDARD_NAME + "," + GLib.FileAttribute.STANDARD_TYPE;
+                    var enumerator = yield file.enumerate_children_async (ATTRIBUTES, GLib.FileQueryInfoFlags.NOFOLLOW_SYMLINKS, GLib.Priority.DEFAULT, cancellable);
+                    GLib.List<GLib.FileInfo> infos;
+                    while ((infos = yield enumerator.next_files_async (1, GLib.Priority.DEFAULT, cancellable)).nth_data (0) != null) {
+                        foreach (unowned GLib.FileInfo info in infos) {
```

Comment:
Since you are reading one file at a time do you need the foreach loop?  Would it be more efficient to read more infos at a time?

Reply:
Yeah probably, I've increased it blindly to 10 as I don't have any benchmark regarding this (and it might differ depending on people hardware…)